File: client/js/updater.js
define(['character', 'timer'], function (Character, Timer) {

  var Updater = Class.extend({
    init: function (game) {
      this.game = game;
      this.playerAggroTimer = new Timer(1000);
    },

    update: function () {
      this.updateZoning();
      this.updateCharacters();
      this.updatePlayerAggro();
      this.updateTransitions();
      this.updateAnimations();
      this.updateAnimatedTiles();
      this.updateChatBubbles();
      this.updateInfos();
      this.updateKeyboardMovement();
    },

    updateCharacters: function () {
      this.game.forEachEntity(function (entity) {
        var isCharacter = entity instanceof Character;

        if (entity.isLoaded) {
          if (isCharacter) {
            this.updateCharacter(entity);
            this.game.updateCharacter(entity);
          }

          this.updateEntityFading(entity);
        }
      }.bind(this));
    },

    updatePlayerAggro: function () {
      var t = this.game.currentTime,
        player = this.game.player;

      // Check player aggro every 1s when not moving nor attacking
      if (player && !player.isMoving() && !player.isAttacking() && this.playerAggroTimer.isOver(t)) {
        player.checkAggro();
      }
    },

    updateEntityFading: function (entity) {
      if (entity && entity.isFading) {
        var duration = 1000,
          t = this.game.currentTime,
          dt = t - entity.startFadingTime;

        if (dt > duration) {
          this.isFading = false;
          entity.fadingAlpha = 1;
        } else {
          entity.fadingAlpha = dt / duration;
        }
      }
    },

    updateTransitions: function () {
      this.game.forEachEntity(function (entity) {
        if (entity.movement && entity.movement.inProgress) {
          entity.movement.step(this.game.currentTime);
        }
      }.bind(this));

      if (this.game.currentZoning && this.game.currentZoning.inProgress) {
        this.game.currentZoning.step(this.game.currentTime);
      }
    },

    updateZoning: function () {
      var g = this.game,
        c = g.camera,
        z = g.currentZoning,
        s = 3,
        ts = 16,
        speed = 500;

      if (z && z.inProgress === false) {
        var orientation = this.game.zoningOrientation,
          startValue = endValue = offset = 0,
          updateFunc = null,
          endFunc = null;

        if (orientation === Types.Orientations.LEFT || orientation === Types.Orientations.RIGHT) {
          offset = (c.gridW - 2) * ts;
          startValue = (orientation === Types.Orientations.LEFT) ? c.x - ts : c.x + ts;
          endValue = (orientation === Types.Orientations.LEFT) ? c.x - offset : c.x + offset;
          updateFunc = function (x) {
            c.setPosition(x, c.y);
            g.initAnimatedTiles();
            g.renderer.renderStaticCanvases();
          }
          endFunc = function () {
            c.setPosition(z.endValue, c.y);
            g.endZoning();
          }
        } else if (orientation === Types.Orientations.UP || orientation === Types.Orientations.DOWN) {
          offset = (c.gridH - 2) * ts;
          startValue = (orientation === Types.Orientations.UP) ? c.y - ts : c.y + ts;
          endValue = (orientation === Types.Orientations.UP) ? c.y - offset : c.y + offset;
          updateFunc = function (y) {
            c.setPosition(c.x, y);
            g.initAnimatedTiles();
            g.renderer.renderStaticCanvases();
          }
          endFunc = function () {
            c.setPosition(c.x, z.endValue);
            g.endZoning();
          }
        }

        z.start(this.game.currentTime, updateFunc, endFunc, startValue, endValue, speed);
      }
    },

    updateCharacter: function (c) {
      // Estimate of the movement distance for one update
      var tick = Math.round(16 / Math.round((c.moveSpeed / (1000 / this.game.renderer.FPS))));

      if (c.isMoving() && c.movement.inProgress === false) {
        if (c.orientation === Types.Orientations.LEFT) {
          c.movement.start(
            this.game.currentTime,
            function (x) {
              c.x = x;
              c.moved();
            },
            function () {
              c.x = c.movement.endValue;
              c.moved();
              c.nextStep();
            },
            c.x - tick,
            c.x - 16,
            c.moveSpeed
          );
        } else if (c.orientation === Types.Orientations.RIGHT) {
          c.movement.start(
            this.game.currentTime,
            function (x) {
              c.x = x;
              c.moved();
            },
            function () {
              c.x = c.movement.endValue;
              c.moved();
              c.nextStep();
            },
            c.x + tick,
            c.x + 16,
            c.moveSpeed
          );
        } else if (c.orientation === Types.Orientations.UP) {
          c.movement.start(
            this.game.currentTime,
            function (y) {
              c.y = y;
              c.moved();
            },
            function () {
              c.y = c.movement.endValue;
              c.moved();
              c.nextStep();
            },
            c.y - tick,
            c.y - 16,
            c.moveSpeed
          );
        } else if (c.orientation === Types.Orientations.DOWN) {
          c.movement.start(
            this.game.currentTime,
            function (y) {
              c.y = y;
              c.moved();
            },
            function () {
              c.y = c.movement.endValue;
              c.moved();
              c.nextStep();
            },
            c.y + tick,
            c.y + 16,
            c.moveSpeed
          );
        }
      }
    },

    updateAnimations: function () {
      var t = this.game.currentTime;

      this.game.forEachEntity(function (entity) {
        if (entity.currentAnimation && entity.currentAnimation.update(t)) {
          entity.dirty();
        }
      });

      if (this.game.sparksAnimation) {
        this.game.sparksAnimation.update(t);
      }

      if (this.game.targetAnimation) {
        this.game.targetAnimation.update(t);
      }
    },

    updateAnimatedTiles: function () {
      var t = this.game.currentTime;

      this.game.forEachAnimatedTile(function (tile) {
        if (!tile.animate(t)) {
          return;
        }
        
        tile.isDirty = true;
        tile.dirtyRect = this.game.renderer.getTileBoundingRect(tile);

        if (this.game.renderer.mobile || this.game.renderer.tablet) {
          this.game.checkOtherDirtyRects(tile.dirtyRect, tile, tile.x, tile.y);
        }
      }.bind(this));
    },

    updateChatBubbles: function () {
      var t = this.game.currentTime;

      this.game.bubbleManager.update(t);
    },

    updateInfos: function () {
      var t = this.game.currentTime;

      this.game.infoManager.update(t);
    },

    updateKeyboardMovement: function () {
      if (!this.game.player || this.game.player.isMoving()) return;

      this.game.selectedCellVisible = false;

      var game = this.game;
      var player = this.game.player;

      var pos = {
        x: player.gridX,
        y: player.gridY
      };

      if (player.moveUp) {
        pos.y -= 1;
        game.keys(pos, Types.Orientations.UP);
      } else if (player.moveDown) {
        pos.y += 1;
        game.keys(pos, Types.Orientations.DOWN);
      } else if (player.moveRight) {
        pos.x += 1;
        game.keys(pos, Types.Orientations.RIGHT);
      } else if (player.moveLeft) {
        pos.x -= 1;
        game.keys(pos, Types.Orientations.LEFT);
      }
    }
  });

  return Updater;
});

File: client/dart/updater.dart
library updater;

import "dart:async";

import "animatedtile.dart";
import "base.dart";
import "character.dart";
import "entity.dart";
import "game.dart";
import "position.dart";
import "lib/gametypes.dart";

class Updater extends Base {

  void update() {
    int t = Game.currentTime;

    this.updateZoning(t);
    this.updateCharacters(t);
    this.updateTransitions(t);
    this.updateAnimations(t);
    this.updateAnimatedTiles(t);
    this.updateChatBubbles(t);
    this.updateInfos(t);
    this.updateKeyboardMovement(t);
  }

  void updateCharacters(int t) {
    Game.forEachEntity((Entity entity) {
      if (!entity.isLoaded) {
        return;
      }

      if (entity is Character) {
        this.updateCharacter(entity);
        Game.updateCharacter(entity);
      }

      this.updateEntityFading(entity);
    });
  }

  void updateEntityFading(Entity entity) {
    if (!entity.isFading) {
      return;
    }

    int duration = 1000;
    int dt = Game.currentTime - entity.startFadingTime;

    if (dt > duration) {
      entity.isFading = false;
      entity.fadingAlpha = 1;
    } else {
      entity.fadingAlpha = dt / duration;
    }
  }

  void updateTransitions(int t) {
    Game.forEachCharacter((Character character) {
      if (character.movement.inProgress) {
        character.movement.step(t);
      }
    });

    if (Game.currentZoning != null && Game.currentZoning.inProgress) {
      Game.currentZoning.step(t);
    }
  }

  void updateZoning(int t) {
    int ts = 16;
    int speed = 500;

    if (Game.currentZoning == null || Game.currentZoning.inProgress) {
      return;
    }

    Orientation orientation = Game.zoningOrientation;
    int startValue = 0;
    int endValue = 0;
    int offset = 0;
    Function updateFunc;
    Function endFunc;

    if (orientation == Orientation.LEFT || orientation == Orientation.RIGHT) {
      offset = (Game.camera.gridW - 2) * ts;
      startValue = (orientation == Orientation.LEFT) ? Game.camera.x - ts : Game.camera.x + ts;
      endValue = (orientation == Orientation.LEFT) ? Game.camera.x - offset : Game.camera.x + offset;
      updateFunc = (x) {
        Game.camera.x = x;
        Game.initAnimatedTiles();
        Game.renderer.renderStaticCanvases();
      };
      endFunc = () {
	Game.camera.x = Game.currentZoning.endValue;
        Game.endZoning();
      };
    } else if (orientation == Orientation.UP || orientation == Orientation.DOWN) {
      offset = (Game.camera.gridH - 2) * ts;
      startValue = (orientation == Orientation.UP) ? Game.camera.y - ts : Game.camera.y + ts;
      endValue = (orientation == Orientation.UP) ? Game.camera.y - offset : Game.camera.y + offset;
      updateFunc = (y) {
	Game.camera.y = y;
        Game.initAnimatedTiles();
        Game.renderer.renderStaticCanvases();
      };
      endFunc = () {
	Game.camera.y = Game.currentZoning.endValue;
        Game.endZoning();
      };
    }

    Game.currentZoning.start(t, updateFunc, endFunc, startValue, endValue, speed);
  }

  void updateCharacter(Character c) {
    // Estimate of the movement distance for one update
    num tick = (16 / ((c.moveSpeed / (1000 / Game.renderer.FPS))).round()).round();

    if (c.isMoving() && c.movement.inProgress == false) {
      if (c.orientation == Orientation.LEFT) {
        c.movement.start(
          Game.currentTime,
          (x) {
            c.x = x;
            c.moved();
          },
          () {
            c.x = c.movement.endValue;
            c.moved();
            c.nextStep();
          },
          c.x - tick,
          c.x - 16,
          c.moveSpeed
        );
      } else if (c.orientation == Orientation.RIGHT) {
        c.movement.start(
          Game.currentTime,
          (x) {
            c.x = x;
            c.moved();
          },
          () {
            c.x = c.movement.endValue;
            c.moved();
            c.nextStep();
          },
          c.x + tick,
          c.x + 16,
          c.moveSpeed
        );
      } else if (c.orientation == Orientation.UP) {
        c.movement.start(
          Game.currentTime,
          (y) {
            c.y = y;
            c.moved();
          },
          () {
            c.y = c.movement.endValue;
            c.moved();
            c.nextStep();
          },
          c.y - tick,
          c.y - 16,
          c.moveSpeed
        );
      } else if (c.orientation == Orientation.DOWN) {
        c.movement.start(
          Game.currentTime,
          (y) {
            c.y = y;
            c.moved();
          },
          () {
            c.y = c.movement.endValue;
            c.moved();
            c.nextStep();
          },
          c.y + tick,
          c.y + 16,
          c.moveSpeed
        );
      }
    }
  }

  void updateAnimations(int t) {
    Game.forEachEntity((Entity entity) {
      if (entity.currentAnimation != null && entity.currentAnimation.update(t)) {
        entity.dirty();
      }
    });

    if (Game.sparksAnimation != null) {
      Game.sparksAnimation.update(t);
    }

    if (Game.targetAnimation != null) {
      Game.targetAnimation.update(t);
    }
  }

  void updateAnimatedTiles(int t) {
    Game.forEachAnimatedTile((AnimatedTile tile) {
      tile.update(t);
    });
  }

  void updateChatBubbles(int t) {
    Game.bubbleManager.update(t);
  }

  void updateInfos(int t) {
    Game.infoManager.update(t);
  }

  void updateKeyboardMovement(int t) {
    if (Game.player == null || Game.player.isMoving()) {
      return;
    }

    Game.selectedCellVisible = false;

    Position pos = Game.player.gridPosition;
    List<Orientation> directions = Game.player.directions.toList();
    Orientation direction = directions.isEmpty ? null : directions.last; 

    switch (direction) {
      case Orientation.UP:
        Game.keys(pos.decY(), Orientation.UP);
        break;

      case Orientation.DOWN:
        Game.keys(pos.incY(), Orientation.DOWN);
        break;

      case Orientation.RIGHT:
        Game.keys(pos.incX(), Orientation.RIGHT);
        break;

      case Orientation.LEFT:
        Game.keys(pos.decX(), Orientation.LEFT);
        break;
    }
  }
}
